File: src/main/java/org/gd/leetcode/p0014/Solution.java
package org.gd.leetcode.p0014;

import org.gd.leetcode.common.LeetCode;

/**
 * https://leetcode.com/problems/longest-common-prefix/
 *
 * @author Gorkhover D.
 * @since 2018-10-21
 */
@LeetCode(
        name = "Longest Common Prefix",
        state = LeetCode.State.DONE,
        difficulty = LeetCode.Level.EASY,
        tags = LeetCode.Tags.STRING
)
class Solution {

    private static String findPrefix(String str0, String str1) {
        int i0 = 0, l0 = str0.length();
        int i1 = 0, l1 = str1.length();

        for (; i0 < l0 && i1 < l1; i0++, i1++) {
            if (str0.charAt(i0) != str1.charAt(i1))
                break;
        }

        if (i0 == l0) return str0;
        if (i1 == l1) return str1;

        return str0.substring(0, i0);
    }

    public String longestCommonPrefix(String[] strings) {
        if (strings == null || strings.length == 0)
            return "";

        String prefix = strings[0];
        if (strings.length == 1 || prefix.isEmpty())
            return prefix;

        for (int i = 1; i < strings.length; i++) {
            prefix = findPrefix(prefix, strings[i]);
            if (prefix.isEmpty())
                return "";
        }

        return prefix;
    }
}

File: src/main/java/org/gd/leetcode/p0014/solution.py
from typing import List


class Solution:

    def longestCommonPrefix(self, strs: List[str]) -> str:

        if not str or len(strs) == 0:
            return ""

        __prefix: str = strs[0]
        if len(strs) == 1 or len(__prefix) == 0:
            return __prefix

        def prefix(str0: str, str1: str) -> str:
            l0 = len(str0)
            l1 = len(str1)
            i0 = 0
            i1 = 0
            while i0 < l0 and i1 < l1:
                if str0[i0] != str1[i1]:
                    break
                i0 += 1
                i1 += 1

            if i0 == l0:
                return str0
            if i1 == l1:
                return str1

            return str[:i0]

        i = 1
        n = len(strs)
        while i < n:
            __prefix = prefix(__prefix, strs[i])
            if len(__prefix) == 0:
                return ""
            i += 1

        return __prefix

